### ISSUE TITLE
Deadlock between test task execution and dependency resolution

### ISSUE THREAD

**Issue by ghale** [gradle/gradle#20269]
With dd8ac8a9276e8d16bcb14dbb9074c79614618e10, when the test task is waiting for tests to execute, it releases the worker lease, but retains the project lock.  This now creates a deadlock scenario where if another task tries to resolve an included build dependency on the same project, it will block waiting on the project lock.  So you end up with the test task holding the project lock, waiting to reacquire a worker lease, and a dependency resolution thread holding a worker lease, waiting to acquire the project lock.  If this happens enough times, it can fill up max-workers with these blocked threads.

There are really three issues here that all conspire to create the deadlock:
- The test task holds the project lock while waiting on the worker lease (this is new)
- The dependency resolution thread holds the worker lease while waiting on the project lock (this has been around for a while)
- The resolved node in the included build is thrown away if the configuration is not used by anything else in the included build, causing the input snapshotting of tasks outside of the included build to re-resolve the configuration (this has also been around for a while)

Fixing any one of those things _should_ remove the possibility of the deadlock.

### Expected Behavior
The build does not deadlock.

### Current Behavior
The build deadlocks.

### Context
This came up in a composite with around 200 included builds.  Because of the high number of included builds, there are lots of opportunity for this issue to surface.
